kolibri/core/auth/management/commands/bulkimportusers.py

```diff
@@ -0,0 +1,614 @@
+import csv
+import logging
+import re
+import sys
+
+from django.contrib.auth.hashers import make_password
+from django.core.exceptions import ValidationError
+from django.core.management.base import CommandError
+
+from kolibri.core.auth.constants.demographics import choices
+from kolibri.core.auth.csv_utils import input_fields
+from kolibri.core.auth.csv_utils import labels
+from kolibri.core.auth.models import Classroom
+from kolibri.core.auth.models import Facility
+from kolibri.core.auth.models import FacilityUser
+from kolibri.core.auth.models import Membership
+from kolibri.core.tasks.management.commands.base import AsyncCommand
+from kolibri.core.tasks.utils import get_current_job
+
+try:
+    FileNotFoundError
+except NameError:
+    FileNotFoundError = IOError
+
+logger = logging.getLogger(__name__)
+DEFAULT_PASSWORD = make_password("kolibri")
+fieldnames = (
+    "Username",
+    "Password",
+    "Full name",
+    "User type",
+    "Identifier",
+    "Birth year",
+    "Gender",
+    "Enrolled in",
+    "Assigned to",
+)
+roles_map = {
+    "LEARNER": None,
+    "ADMIN": "admin",
+    "FACILITY COACH": "coach",
+    "CLASS COACH": "classroom assignable coach",
+}
+
+# Validators ###
+
+
+def number_range(min, max):
+    """
+    Return a value check function which raises a ValueError if the supplied
+    value is less than `min` or greater than `max`.
+    """
+
+    def checker(v):
+        if int(v) < min or int(v) > max:
+            raise ValueError(v)
+
+    return checker
+
+
+def value_length(length, allow_null=False):
+    """
+    Return a value check function which raises a ValueError if the supplied
+    value has a length greater than 'length'
+    If null is not True raises a ValueError if the supplied value is None.
+    """
+
+    def checker(v):
+        if allow_null and v is None:
+            return checker
+        if len(v) > length:
+            raise ValueError(v)
+
+    return checker
+
+
+def enumeration(*args):
+    """
+    Return a value check function which raises a ValueError if the value (case
+    insensitive) is not in the enumeration of values provided by args.
+    """
+    assert len(args) > 0, "at least one argument is required"
+    if len(args) == 1:
+        # assume the first argument defines the membership
+        members = args[0].lower()
+    else:
+        members = tuple(map(str.lower, args))
+
+    def checker(value):
+        if value.lower() not in members:
+            raise ValueError(value)
+
+    return checker
```

Comment:
This function is a bit confusing because of the `assert` and `if`/`else` branches. It's also only used twice, which seems to imply that it should really just be two functions with more explicit function signatures?

Reply:
ah, you're right, the assert was in my first iteration doing it aka TDD, I'll remove it. I's not needed anymore.
In any case, the if/else are for a different reason, enumeration could be a string (not applicable in our case but I did it as a general validator, in case you want to ensure a char is in, for example "ABCXY")